File: kyonggiyo-api/src/main/java/kyonggiyo/api/adapter/controller/user/UserController.java
package kyonggiyo.api.adapter.controller.user;

import kyonggiyo.api.adapter.controller.user.dto.UserCreateRequest;
import kyonggiyo.api.adapter.controller.user.dto.ValidateNicknameRequest;
import kyonggiyo.api.adapter.controller.user.dto.ValidateNicknameResponse;
import kyonggiyo.application.port.in.auth.ProvideAuthCodeUrlUseCase;
import kyonggiyo.application.port.in.user.CreateUserUseCase;
import kyonggiyo.application.port.in.user.ValidateNicknameUseCase;
import kyonggiyo.domain.auth.Platform;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.net.URI;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/users/")
public class UserController {

    private final CreateUserUseCase createUserUseCase;
    private final ProvideAuthCodeUrlUseCase provideAuthCodeUrlUseCase;
    private final ValidateNicknameUseCase validateNicknameUseCase;

    @PostMapping("/profile")
    public ResponseEntity<Void> userCreate(@RequestBody UserCreateRequest request) {
        Platform platform = createUserUseCase.createUser(request.toCommand());
        URI uri = provideAuthCodeUrlUseCase.provideUri(platform);
        return ResponseEntity.created(uri).build();
    }

    @GetMapping("/profile/nickname")
    public ResponseEntity<ValidateNicknameResponse> nicknameValidate(ValidateNicknameRequest request) {
        boolean existed = validateNicknameUseCase.existByNickname(request.nickname());
        return ResponseEntity.ok(new ValidateNicknameResponse(!existed));
    }

}

File: kyonggiyo-api/src/main/java/kyonggiyo/api/adapter/controller/user/UserController.kt
package kyonggiyo.api.adapter.controller.user

import kyonggiyo.api.adapter.controller.user.dto.UserCreateRequest
import kyonggiyo.api.adapter.controller.user.dto.UserDeleteRequest
import kyonggiyo.api.adapter.controller.user.dto.ValidateNicknameRequest
import kyonggiyo.api.adapter.controller.user.dto.ValidateNicknameResponse
import kyonggiyo.application.auth.port.inbound.ProvideAuthCodeUrlUseCase
import kyonggiyo.application.port.`in`.user.CreateUserUseCase
import kyonggiyo.application.port.`in`.user.ValidateNicknameUseCase
import kyonggiyo.application.port.`in`.user.WithdrawUserUseCase
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/users/")
class UserController(
        private val createUserUseCase: CreateUserUseCase,
        private val provideAuthCodeUrlUseCase: ProvideAuthCodeUrlUseCase,
        private val validateNicknameUseCase: ValidateNicknameUseCase,
        private val withdrawUserUseCase: WithdrawUserUseCase,
) {

    @PostMapping("/profile")
    fun userCreate(@RequestBody request: UserCreateRequest): ResponseEntity<Unit> {
        return createUserUseCase.createUser(request.toCommand())
            .let { provideAuthCodeUrlUseCase.provideUri(it) }
            .let { ResponseEntity.created(it).build() }
    }

    @GetMapping("/profile/nickname")
    fun nicknameValidate(request: ValidateNicknameRequest): ResponseEntity<ValidateNicknameResponse> {
        return validateNicknameUseCase.existByNickname(request.nickname)
            .let { ResponseEntity.ok(ValidateNicknameResponse(it)) }
    }

    @DeleteMapping("/profile/withdraw")
    fun withdraw(@RequestBody request: UserDeleteRequest): ResponseEntity<Unit> {
        withdrawUserUseCase.deleteUser(request.toCommand())
        return ResponseEntity.noContent().build()
    }
}